dash_labs/dashdown.py

```diff
@@ -0,0 +1,419 @@
+    """
+    certain props may not be updated for security reasons and/or because they are updated by
+    the MarkdownAIO component
+    """
+
+    for prop in props:
+        if isinstance(props[prop], dict):
+            for p in props[prop]:
+                if p in prohibited_props[prop]:
+                    raise Exception(
+                        f" The `{p}` prop in `{prop}` may not be user-supplied in MarkdownAIO"
+                        + f" see code block# {code_block} in module {filename}"
+                        if code_block
+                        else ""
+                    )
+
+
+def _run_code(code, scope=None, div_props=None):
+    if scope is None:
+        scope = {}
+
+    # Replacements to get this working in an `exec` environment:
+    # 1) Remove language specifier included at beginning of markdown string
+    if code.startswith("python"):
+        code = code[6:]
+    # 2) Remove the app instance in the code block otherwise app.callbacks don't work
+    code = _remove_app_instance(code)
+
+    if "app.layout" in code:
+        code = code.replace("app.layout", "layout")
+    if "layout" in code:
+        exec(code, scope)
+        return html.Div(scope["layout"], **div_props)
+
+    else:
+        # taken from https://stackoverflow.com/a/39381428
+        block = ast.parse(code, mode="exec")
+        # assumes last node is an expression
+        last = ast.Expression(block.body.pop().value)
+        exec(compile(block, "<string>", mode="exec"), scope)
+        return html.Div(eval(compile(last, "<string>", mode="eval"), scope))
+
+
+def _parse_props(component_props, section):
+    """
+    returns the component property dict specified within a code block.
+
+    For example if this was on the first line of the code block:
+       `app-div-props-{"style": {"maxHeight": "800px"}}`
+    it would return the dict:
+        `{"style": {"maxHeight": "800px"}}`
+    """
+    section_split = section.split(component_props, 1)[-1]
+    matched_paren = 0
+    parsed_dict = []
+    for i in section_split:
+        parsed_dict.append(i)
+        if i == "{":
+            matched_paren += 1
+        if i == "}":
+            matched_paren -= 1
+        if matched_paren == 0:
+            return ast.literal_eval("".join(parsed_dict))
+
+
+def _update_props(
+    props_dict, prohibited_props, section, code_block=None, filename=None
+):
+    """
+    update props that are user-supplied  within a code block
+    """
+    props_dict = copy.deepcopy(props_dict)
+    if "side-by-side-true" in section:
+        props_dict["side_by_side"] = True
+    if "side-by-side-false" in section:
+        props_dict["side_by_side"] = False
+    if "code-first-true" in section:
+        props_dict["code_first"] = True
+    if "code-first-false" in section:
+        props_dict["code_first"] = False
+    if "exec-code-true" in section:
+        props_dict["exec_code"] = True
+    if "exec-code-false" in section:
+        props_dict["exec_code"] = False
```

Comment:
🙂 This works! I was thinking of something that might've been a bit more generic, like you'd pass in the list of properties into the function `['side_by_side', `code_first', `exec_code']`, replace the `_` with `-`, and then split out the value on the rhs of `=` and `exec` or `eval` it to get the value.

Reply:
Then you could loop through all of the props and the function would be quite a bit shorter and easier to maintain

Reply:
Made an update, but I know it's still ugly :pig: :lipstick:   Will refactor more once we settle on the format of the props in the code block